keps/sig-apps/2804-consolidate-workload-controllers-status/README.md

```diff
@@ -191,12 +222,40 @@ See the [Kubernetes API Conventions](https://git.k8s.io/community/contributors/d
 Because of the number of changes that are involved as part of this effort, we are thinking of a phased approach where we introduce these conditions to DaemonSet controller first and then make similar changes to ReplicaSet and StatefulSet controller. We would graduate ExtendedWorkloadConditions to beta once all the features and `progressDeadlineSeconds` KEP are implemented.
 This also needs some code refactoring of existing conditions for Deployment controller.
 
-#### Available
-Individual workload controllers mark a ReplicaSet or StatefulSet as `available` when number of available replicas reaches number of replicas.
-- This could be confusing in ReplicaSets a bit since Deployment could get available sooner than its ReplicaSet due `maxUnavailable`.
-- Available replicas is alpha feature guarded by StatefulSetMinReadySeconds gate in StatefulSets, but the value defaults to ReadyReplicas when the feature gate is disabled so using it shouldn't be an issue.
+#### Operational
```

Comment:
I think this shouldn't be added by kubernetes.

From https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

> Objects may report multiple conditions, and new types of conditions may be added in the future or by 3rd party controllers.

It doesn't sound like this condition would be useful for a wide number of users. In particular, it's very hard to interpret for Jobs, as opposed to just looking at .status.active. But OTOH users are free to set it up based on their own policies.

Reply:
> I think this shouldn't be added by kubernetes.

> Objects may report multiple conditions, and new types of conditions may be added in the future or by 3rd party controllers.

I am not sure, if this is a sufficient reason. Can you please be more specific in which area this would be problematic? I have looked into cloud native projects and have not found any such use case.

> It doesn't sound like this condition would be useful for a wide number of users. In particular, it's very hard to interpret for Jobs, as opposed to just looking at .status.active. But OTOH users are free to set it up based on their own policies.

I am sorry this was poorly formulated in the beginning (it is hopefully better now). As described in the proposal the `Operational` condition will not be added into Jobs or CronJobs. https://github.com/atiratree/kube-enhancements/blob/consolidate-status/keps/sig-apps/2804-consolidate-workload-controllers-status/README.md#proposed-conditions I think it makes sense for the workloads where this is mentioned. Do you have some remarks to the workloads where it is proposed?

As we discussed before, we are not targeting Jobs and CronJobs at this time. And we would update later our evaluation what to do in that space.

Reply:
My PoV is more general than just Jobs:

If Operational has a fixed policy (for example, all pods are available), it's unlikely to be useful to all users. I also think that having some spec fields to control when the Operational condition is added might be an overkill. I guess the question is what use case is this condition enabling.

What is wrong with just exposing the number of active of workloads and let users interpret that as they wish? Or write their own controllers to append a condition that is meaningful to them.

Reply:
> My PoV is more general than just Jobs:
> 
> If Operational has a fixed policy (for example, all pods are available), it's unlikely to be useful to all users. I also think that having some spec fields to control when the Operational condition is added might be an overkill. 

I agree, as I suggested in the KEP we could first introduce the condition and once there is additional interest consider adding customization via the spec field

we can be more thorough with the feedback and wait for a longer time before graduating this to be sure we are committed to the right thing 

> I guess the question is what use case is this condition enabling.

this is mainly for better user experience , but yes for example operators can do without this feature and use the status fields

cons:

- better synergy with kubectl https://github.com/kubernetes/kubernetes/issues/79606
- as mentioned less dependency on other tools like `kapply status` or custom scripts
- familiarity with conditions across all the workloads 

> What is wrong with just exposing the number of active of workloads and let users interpret that as they wish? Or write their own controllers to append a condition that is meaningful to them.

- inconsistency: workloads expose that via different fields (e.g DS vs STS)
- not everybody knows the details of each status and this condition would give an interface for a common user
-  it is not really feasible for each user to implement their own controllers for this when you consider maintainance, RBAC permissions, collisions with other controllers. I think it makes for this to be implemented by a platform so everybody can benefit from this common functionality

Reply:
Any controller that exposes an update mechanism that defines “availability” as a constraint (minReady, maxUnavailable) by definition is maintaining an Available condition.  Since all three key workload controllers with update strategies expose those fields deliberately, I am not sure why we are stating “Available” is not consistent if added to those workload types.  They were designed to be consistent.  I’ll reread the proposed changes, but I’m a bit confused and hope I’m not missing something fundamental here.

Reply:
Yes, the controllers are maintaining it, but they are not exposing it as such. It is up to the user to interpret the actual values; e.g. `.status.availableReplicas`. We could expose the condition if we wanted to, but I think it might be confusing to the users, as they might not agree with our definition. For example, take a statefulset with 1 replica (that requires maxUnavailable to be 1)- it would mark it as Available even if 0 replicas were running.

Reply:
I don't understand how users can "disagree with our definition"?  Our API has a very clear meaning, we expose fields the user sets to maintain that condition.  So they can't really say "nope, I don't agree that when I set maxUnavailable to 80%, and I reached 21% of instances down, it shouldn't be available".

>  For example, take a statefulset with 1 replica (that requires maxUnavailable to be 1)- it would mark it as Available even if 0 replicas were running.

This would be a bug if it's happening today.  The deployment controller should ALWAYS report `Available: false` in this case.  You can't be available at the fencepost.  Is this happening?

Reply:
Roughly, the way the controllers should operate is:

1. upgrade variables define *how* to make an upgrade, but the controller is accountable for ensuring forward progress (i.e. bad parameters can't block upgrade)
2. Available as a condition is described using those parameters, but with the understanding that *some* upgrade options will favor "breaking availability" over "making no progress because the parameters are impossible"
3. So when you ask for those parameters `maxUnavailable: 0, maxSurge:0` and start an upgrade if you only have one replica your condition is `Available:false` when the controller stops that instance, and that as intended (if it's not, it's a bug!)

Reply:
> This would be a bug if it's happening today. The deployment controller should ALWAYS report Available: false in this case. You can't be available at the fencepost. Is this happening?

Yes this is happening in Deployments today. Nit: (`maxUnavailable: 1, maxSurge:0`)
-  we can discuss if we should fix it, since this has been the default behaviour till now (I can bring this up in Monday's sig-apps call) 
- the problem of unavailability can be circumvented with `maxSurge` in Deployments, but there is no such an option for StatefulSets, and as you have pointed out we have to accept the unavailability when doing a rollout.


> I don't understand how users can "disagree with our definition"? Our API has a very clear meaning, we expose fields the user sets to maintain that condition. So they can't really say "nope, I don't agree that when I set maxUnavailable to 80%, and I reached 21% of instances down, it shouldn't be available".

> 1. upgrade variables define how to make an upgrade, but the controller is accountable for ensuring forward progress (i.e. bad parameters can't block upgrade)
> 2. Available as a condition is described using those parameters, but with the understanding that some upgrade options will favor "breaking availability" over "making no progress because the parameters are impossible"

I agree about the upgrade part, but we are a bit ambiguous on the API definition part
- it defines [maxUnavailable](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#max-unavailable) only  with respect to the rollout/update/upgrade.
- we mention in [types](https://github.com/kubernetes/kubernetes/blob/77bf2d1ade933ad5f7c78216cb4729f2cefd0e55/pkg/apis/apps/types.go#L522-L523) and imply (by reconciling) that this is checking availability of the application even after the rollout with the rollout constraints. Example thought process of how users might look at this: [blog post](https://maelvls.dev/deployment-available-condition/) 

The workloads user cannot specify a different value after the rollout to tune it specifically for their application. For example the user could want fast upgrades/rollouts and accept many replicas down in that time (maybe even a downtime), but demand only a few (or none) replicas down in day to day operations

I think the `Operational` condition could help achieving consistent monitoring for users of all workloads, regardless of a rollout and type of the workload. Also our state in a successful rollout is  100% of replicas available (`Progressing` > `NewReplicaSetAvailable`), so I think it could make sense to implement `Operational` with 100% as well, and iterate on that (allow customization) once we get a feedback. The advantage is we can also introduce this to non rollout workloads (ReplicaSets) and other strategies (Recreate, OnDelete).

Nevertheless, we could still introduce `Available` for rollout workloads to express how the replicas are doing during a rollout, but we should take the following into the account to see if it is worth pursuing:
- if we fix the bug, the condition will be defined in a more complicated way (but less surprising way in some sense) and a bit harder to use from monitoring perspective, since you can expect the valid states to be both `True` and `False` during a rollout.
- we are trying to cover mainly the upgrade part but the users expect the condition to represent availability / healthiness of the app. We should be clear on the definition and the use cases because the name `Available` is very broad

Reply:
> The workloads user cannot specify a different value after the rollout to tune it specifically for their application.

That is a not unreasonable requirement we could add to all workload controllers.

Reply:
https://github.com/kubernetes/kubernetes/blob/fce22c807f2851a26811332df9952a1826e6dcc1/pkg/controller/deployment/sync.go#L514

Has a fencepost bug.

```
	if availableReplicas >= *(deployment.Spec.Replicas)-deploymentutil.MaxUnavailable(*deployment) {
```

should be 

```
        if availableReplicas > 0 && ...
```

You can't be available with zero pods.  What user behavior would be dependent on this?

Reply:
Opened https://github.com/kubernetes/kubernetes/pull/113228 - if there is a backwards compatibility issue we can discuss it there.

Reply:
Ok, going through this I see the issue is that we *are* consistent across both, we just defined it in a way that makes sense for machines (apply the rule consistently) but not necessarily for humans (not having replicas is not available at a deployment level).  I agree with the assessment that we cannot change the behavior of the condition because of that consistency.

Given that, we SHOULD still implement Available consistently across all controllers.  Being inconsistent between types is just as bad as being inconsistent within a given type (if available was reported differently during update and steady state) - we prioritize consistency of the system over net new features.

For that reason I am against introducing a net new condition without first (or at the same time) making Available consistent across the standard types which have rolling upgrade and minimum availability defined.  Will make that comment in the other places.

Reply:
Of course, as I soon as I type this I realize that Recreate deployments *do* go unavailable during update, which doesn't make sense.  The presence of the Recreate strategy could be interpreted to be identical to `replicas=1,rolling,maxUnavailable=25%(default)` - minimum availability of a Recreate deployment is 0, not 100%.  Will need to ponder that tomorrow.